File: agents/dotnet/src/ECommerceAgents.Shared/Tools/ProductLookupTools.cs
using Dapper;
using ECommerceAgents.Shared.Data;
using Microsoft.Extensions.AI;
using System.ComponentModel;

namespace ECommerceAgents.Shared.Tools;

/// <summary>
/// Resolves a product name to its id — the .NET twin of Python's
/// <c>shared/tools/product_lookup_tools.py</c>.
///
/// This is the first entry in a shared .NET tool library (issue #18). It is
/// first because without it the specialists that take a <c>productId</c> are
/// unusable from a normal question: asked "what are customers saying about the
/// Dyson V15?", the model has no way to turn that name into an id, so it
/// invents one. Observed live against the real stack before this shipped —
/// review-sentiment called AnalyzeSentiment, GetSentimentTrend and
/// DetectFakeReviews in turn, each with the made-up id "dyson-v15-id", got
/// "Product not found" three times, and told the user the product doesn't
/// exist. It does exist; only the lookup was missing.
///
/// Attached to the same three specialists Python attaches it to
/// (pricing-promotions, review-sentiment, inventory-fulfillment) — the ones
/// whose tools are keyed by product id. product-discovery and order-management
/// don't need it: they already resolve products by name themselves.
/// </summary>
public sealed class ProductLookupTools(DatabasePool pool)
{
    private readonly DatabasePool _pool = pool;

    public IEnumerable<AITool> All() => new AITool[]
    {
        AgentTool.Create(FindProductByName, nameof(FindProductByName)),
    };

    [Description(
        "Resolve a product's UUID from its name (or a close match). Call this first "
        + "whenever the user refers to a product by name rather than a UUID, before "
        + "calling any tool that requires product_id."
    )]
    public async Task<object> FindProductByName(
        [Description("Product name or a close match, e.g. 'Sony WH-1000XM5'")] string name
    )
    {
        await using var conn = await _pool.OpenAsync();

        // Exact (case-insensitive) match first, then a substring match on every
        // word — the same word-splitting search_products uses, so "Sony
        // headphones" still finds "Sony WH-1000XM5".
        var exact = await conn.QueryFirstOrDefaultAsync<ProductRow>(
            "SELECT id, name FROM products WHERE is_active = TRUE AND name ILIKE @name LIMIT 1",
            new { name }
        );
        if (exact is not null)
        {
            return new { found = true, product_id = exact.Id.ToString(), product_name = exact.Name };
        }

        var words = (name ?? string.Empty)
            .Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Where(w => w.Length >= 2)
            .ToList();

        if (words.Count == 0)
        {
            return new { found = false, message = $"No product matching '{name}'" };
        }

        var conditions = string.Join(" AND ", words.Select((_, i) => $"name ILIKE @w{i}"));
        var args = new DynamicParameters();
        for (var i = 0; i < words.Count; i++)
        {
            args.Add($"w{i}", $"%{words[i]}%");
        }

        var row = await conn.QueryFirstOrDefaultAsync<ProductRow>(
            $"SELECT id, name FROM products WHERE is_active = TRUE AND {conditions} LIMIT 1",
            args
        );

        return row is null
            ? new { found = false, message = $"No product matching '{name}'" }
            : new { found = true, product_id = row.Id.ToString(), product_name = row.Name };
    }

    private sealed record ProductRow(Guid Id, string Name);
}

File: agents/python/shared/tools/product_lookup_tools.py
"""专业智能体共用的商品名称到标识解析。

评论、定价和库存工具需要商品 UUID，但用户通常只提供名称。
提供真实查询能力，避免专业智能体在没有商品搜索工具时编造 UUID，
再把查询空结果误报为商品没有数据。
"""

from __future__ import annotations

from typing import Annotated

from agent_framework import tool
from pydantic import Field

from shared.db import get_pool


@tool(
    name="find_product_by_name",
    description=(
        "Resolve a product's UUID from its name (or a close match). Call this first "
        "whenever the user refers to a product by name rather than a UUID, before "
        "calling any tool that requires product_id."
    ),
)
async def find_product_by_name(
    name: Annotated[str, Field(description="Product name or a close match, e.g. 'Sony WH-1000XM5'")],
) -> dict:
    pool = get_pool()
    async with pool.acquire() as conn:
        # 先忽略大小写做完整名称匹配，再按词做子串匹配，
        # 与 search_products 的分词方式一致，
        # 让品牌与品类组合也能找到具体型号。
        exact = await conn.fetchrow(
            "SELECT id, name FROM products WHERE is_active = TRUE AND name ILIKE $1 LIMIT 1",
            name,
        )
        if exact:
            return {"found": True, "product_id": str(exact["id"]), "product_name": exact["name"]}

        words = [w for w in name.strip().split() if len(w) >= 2]
        if not words:
            return {"found": False, "message": f"No product matching '{name}'"}

        conditions = " AND ".join(f"name ILIKE ${i + 1}" for i in range(len(words)))
        args = [f"%{w}%" for w in words]
        row = await conn.fetchrow(
            f"SELECT id, name FROM products WHERE is_active = TRUE AND {conditions} LIMIT 1",
            *args,
        )
        if not row:
            return {"found": False, "message": f"No product matching '{name}'"}

        return {"found": True, "product_id": str(row["id"]), "product_name": row["name"]}
